internal/pkg/deploy/cloudformation/cloudformation.go

```diff
@@ -374,89 +375,116 @@ func (cf CloudFormation) executeAndRenderChangeSet(in *executeAndRenderChangeSet
 	g, ctx := errgroup.WithContext(context.Background())
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
+	prevStackEventsErased := make(chan bool)
 	g.Go(func() error {
 		defer cancel()
-		if err := cf.renderChangeSet(ctx, changeSetID, in); err != nil {
+		nl, err := cf.renderChangeSet(ctx, changeSetID, in)
+		if err != nil {
 			if !errors.Is(err, context.Canceled) {
 				return err
 			}
+			// Erase previous stack events only if context is canceled
+			// by waitForSignalAndHandleInterrupt().
+			cursor.EraseLinesAbove(cf.console, nl)
+			close(prevStackEventsErased)
 		}
 		return nil
 	})
 	if in.enableInterrupt {
 		g.Go(func() error {
-			return cf.waitForSignalAndHandleInterrupt(ctx, cancel, sigChannel, in.stackName)
+			return cf.waitForSignalAndHandleInterrupt(signalHandlerInput{
+				ctx:                   ctx,
+				cancelFn:              cancel,
+				sigCh:                 sigChannel,
+				stackName:             in.stackName,
+				prevStackEventsErased: prevStackEventsErased,
+			})
 		})
 	}
 	return g.Wait()
 }
 
-func (cf CloudFormation) renderChangeSet(ctx context.Context, changeSetID string, in *executeAndRenderChangeSetInput) error {
+// renderChangeSet renders and executes a CloudFormation change set, providing progress updates if necessary.
+// It returns the number of rendered lines and any encountered error.
+func (cf CloudFormation) renderChangeSet(ctx context.Context, changeSetID string, in *executeAndRenderChangeSetInput) (int, error) {
 	if _, ok := cf.console.(*discardFile); ok { // If we don't have to render skip the additional network calls.
-		return nil
+		return 0, nil
 	}
 	waitCtx, cancelWait := context.WithTimeout(ctx, waitForStackTimeout)
 	defer cancelWait()
 	g, ctx := errgroup.WithContext(waitCtx)
 
 	renderer, err := cf.createChangeSetRenderer(g, ctx, changeSetID, in.stackName, in.stackDescription, progress.RenderOptions{})
 	if err != nil {
-		return err
+		return 0, err
 	}
+	var prevNumLines int
 	g.Go(func() error {
-		_, err := progress.Render(ctx, progress.NewTabbedFileWriter(cf.console), renderer)
+		var err error
+		prevNumLines, err = progress.Render(ctx, progress.NewTabbedFileWriter(cf.console), renderer)
 		return err
 	})
 	if err := g.Wait(); err != nil {
-		return err
+		return prevNumLines, err
 	}
 	if err := cf.errOnFailedStack(in.stackName); err != nil {
-		return err
+		return 0, err
```

Comment:
Should this and line 433 be `return prevNumLines, err` and `return prevNumLines, nil` respectively?